lib/os/sem.c

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <misc/sem.h>
+#include <syscall_handler.h>
+
+void sys_sem_init(struct sys_sem *sem)
+{
+	k_futex_init(&sem->futex);
+}
+
+int sys_sem_give(struct sys_sem *sem)
+{
+	int ret = 0;
+
+	atomic_add(&sem->futex.val, 1);
+	ret = k_futex_wake(&sem->futex, false);
```

Comment:
Currently the code considered SMP case, the multiple operations(more than one instruction) on futex should only happen on one side, take or give, because of contend even it's atomic type. So on give side,  I do k_futex_wake unconditionally. If we don't consider SMP, I can add the compare.

Reply:
Why can't we do:

```
if (atomic_add(&sem->futex.val, 1) == 0) {
   /* Futex was contended. Wake one of the sleeping threads up. */
   ret = k_futex_wake(&sem->futex, false);
}
```

Reply:
This may be worth reviewing:

https://github.com/aosp-mirror/platform_bionic/blob/master/libc/bionic/semaphore.cpp

Reply:
> Why can't we do:
> 
> ```
> if (atomic_add(&sem->futex.val, 1) == 0) {
>    /* Futex was contended. Wake one of the sleeping threads up. */
>    ret = k_futex_wake(&sem->futex, false);
> }
> ```
yes, you are right, I consider more for sys_sem_give.